design-proposals/export-import/vm-export-import.md

```diff
@@ -0,0 +1,166 @@
+# Overview
+Proposal for exporting/importing Kubevirt Virtual Machine data from/to a Kubernetes cluster.
+
+## Motivation
+The Containerized Data Importer (CDI) project is mostly focused on getting Virtual Machine data into a cluster. The DataVolume API declares import and upload operations that populate and initialize individual Persistent Volume Claims (PVCs) with external data.  Of course that data is very likely to change once the PVCs are mounted and consumed by VirtualMachines (VMs).  But there is no standard way to get that valuable data out of the cluster.  Maybe an external backup would be useful for Disaster Recovery purposes?  Or for migrating VMs between clusters?
+
+Once there is a standard API for exporting all persistent data for a VM, it follows that a new type of import operation is required, one that operates on a higher level than the existing CDI primitatives.  It should be able to parse an exported arhive and restore the Virtual Machine configuration and data.
+
+## Goals
+* Define Virtuam Machine Archive file format
+* Define VirtualMachineExport API
+* Introduce virtctl export|import vm commands
+
+## Non Goals
+* Implementation details of internal components (servers/proxies/controllers/etc)
+
+## User Stories
+* As a KubeVirt user, I would like to backup the current state of a Virtual Machine by exporting a Virtual Machine Archive
+* As a KubeVirt user, I would like to backup the state of a Virtual Machine Snapshot by exporting a Virtual Machine Archive
+* As a KubeVirt user, I would like to create a new Virtual Machine from a Virtual Machine Archive
+
+## Repos
+* [KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+## Virtual Machine Archive File Format
+
+A Vritual Machine Archive is a xz compressed tar archive with the following structure
+
+```
+.
+├── manifest
+├── resources
+│   ├── pvc-block-pvc.yaml
+│   ├── pvc-filesystem-pvc.yaml
+│   ├── pvc-virtiofs-pvc.yaml
+│   └── vm.yaml
+└── volumes
+    ├── block-pvc
+    │   └── disk.img
+    ├── filesystem-pvc
+    │   └── disk.img
+    └── virtiofs-pvc
+        ├── file1
+        ├── file2
+        └── file3
+```
+
+### Manifest
+
+The `manifest` file lists all files (excluding itself) in the archive and their SHA-1 hashes.  It can be used to ensure that corrumted data is not imported.
```

Comment:
I don't think we should wed too tightly to a specific cryptographic hash, as any specific one may be [found to have weaknesses](https://www.schneier.com/blog/archives/2005/02/cryptanalysis_o.html). I'd prefer to define which hashes are allowed, and have the manifest state which one of them is used.

I suppose that the manifest should also list the version of the VM archive - we may want to augment or fix bugs in what we define here.

Reply:
Really just trying to protect from bit rot here not active attackers.  If we want to really be cryptographically secure, the thing to do would be to sign the manifest file and include the signature in the archive.  Without a signature file, anyone can tamper with the archive at rest and update the manifest to have the new hash regardless of the hash function.

Reply:
Since we do not need cryptographic powers, CRC may be a more fitting hash. A down side is that xz or container images already have their own checksum, so the point of a per-file hash is less clear. My original comment is to have the algorithm specified in file (or by the export bundle version) so we can change it if sha-1 falls out of grace and anyone using it are flagged as a security breech (as already happened to md5)

BTW I'm not sure about the value of the manifest: tar and container images has their own list of files, so we risk data duplication.

Reply:
Yeah, the manifest/signature files are taken from OVA.  If we decide to have the native export format to be "container" then we may not need them.

Reply:
fwiw, I see value in exporting to a removable tarball as well as "container".

Reply:
Yeah I'm leaning towards that as well